crates/ra_ide_api/src/change.rs

```diff
@@ -243,4 +243,65 @@ impl RootDatabase {
         self.query(hir::db::InferQuery).sweep(sweep);
         self.query(hir::db::BodyHirQuery).sweep(sweep);
     }
+
+    pub(crate) fn per_query_memory_usage(&mut self) -> Vec<(String, Bytes)> {
+        let mut acc: Vec<(String, Bytes)> = vec![];
+        let sweep = SweepStrategy::default().discard_values().sweep_all_revisions();
+        macro_rules! sweep_each_query {
+            ($($q:path)*) => {$(
+                let before = memory_usage().allocated;
+                self.query($q).sweep(sweep);
+                let after = memory_usage().allocated;
+                let q: $q = Default::default();
+                let name = format!("{:?}", q);
+                acc.push((name, before - after));
+            )*}
+        }
+        sweep_each_query![
+            ra_db::ParseQuery
+            ra_db::SourceRootCratesQuery
+            hir::db::AstIdMapQuery
+            hir::db::ParseMacroQuery
+            hir::db::MacroDefQuery
+            hir::db::MacroArgQuery
+            hir::db::MacroExpandQuery
+            hir::db::StructDataQuery
+            hir::db::EnumDataQuery
+            hir::db::TraitDataQuery
+            hir::db::TraitItemsIndexQuery
+            hir::db::RawItemsQuery
+            hir::db::RawItemsWithSourceMapQuery
+            hir::db::CrateDefMapQuery
+            hir::db::ImplsInModuleQuery
+            hir::db::ImplsInModuleWithSourceMapQuery
```

Comment:
wouldn't it make more sense to put the `WithSourceMap` query first? then the source map would (hopefully) get collected with that, and the e.g. `ModuleImplBlocks` would get collected with the other, so we'd get separate numbers (the list you posted in Zulip shows 0kb for `ImplsInModuleQuery` and `RawItemsQuery`)

Reply:
Yeah, makes sense, thanks!